### PR TITLE
[None][feat] k3 weight pipeline opt

### PR DESCRIPTION
**Pull request by WeiHaocheng**
This PR refactors the Kimi K3 text model's weight pipeline in modeling_kimi_linear.py in two commits. The
  first encapsulates the KDA speculative-decode replay-cache state behind a has_kda_replay_caches property
  and a commit_conv_window method on the mamba cache (replacing ad-hoc getattr probing), and renames the
  ambiguous generation counters (num_decodes/decode_rows → num_generations/num_gen_tokens) for clarity. The
  second replaces ~120 lines of hand-written tensor-slicing in load_param with the framework's
  load_weight_shard primitive plus a small _pad_dim helper for the 96→128 head padding the primitive can't
  express — covering the KDA head-shard, MLA head pad-and-shard, shared-expert/dense-MLP TP, and fused
  gate/up paths, all behavior-preserving (net −60 lines). Both changes are compiled, unit-tested, and
  end-to-end validated on the real checkpoint with a pure-TP16 text smoke that exercises the head-shard load
  paths.

## Dev Engineer Review

- Refactors KDA replay-cache handling through `has_kda_replay_caches` and `commit_conv_window`.
- Simplifies weight loading with `load_weight_shard` and `_pad_dim`.
- Updates KDA, MLA, MLP, shared-expert, and fused gate/up sharding paths.
- No configuration or test-list changes were identified.
- The implementation is consistent with the stated objectives.
- CI main L0 pipelines failed across three runs. Investigate these failures before rerunning CI.
- Verdict: needs follow-up.

## QA Engineer Review

- Modified KDA parity coverage in:
  - `test_kimi_kda_fused_verify_parity.py`
  - `test_kimi_kda_verify_parity.py`
- Updated fused and sequential cache setup.
- Updated prefill and multistream decode parity checks for `has_kda_replay_caches`.
- No `tests/integration/test_lists/` coverage information was provided.
- CBTS coverage data is unavailable.
- Verdict: needs follow-up.

### CODE REVIEW

**Comment by nvpohanh** [comment]
[by Codex] @VALLIS-NERIA Could you review this PR? Thanks!

**Review by VALLIS-NERIA — APPROVED** [review]
My approval covers only the KV cache manager portion of this PR.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 1]
```diff
@@ -2567,25 +2563,33 @@ def _load_trunk_params(
                 kda_tp_rank = layer.self_attn._kda_tp_rank
                 break
 
+        COL = TensorParallelMode.COLUMN   # shards output rows (dim 0)
```
Comment (litaotju) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py]: Do we actually use 2-D, column and row for TP?

**Review by litaotju — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py (hunk 1 above)
Comment (WeiHaocheng) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py]: yes, both are used, on 2-D weight matrices. In the current code:                                 

-   ROW (shard input/contraction columns, dim 1): o_proj (KDA) and down_proj (shared-expert / dense MLP)                                                                                        

  - COLUMN (shard output rows, dim 0): fused gate_up (2591–2592), KDA q/k/v/g/f_b/b projections, and        
  shared/MLP gate/up .

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py [hunk 2]
```diff
@@ -17,8 +17,7 @@
 import os
 from abc import ABC, abstractmethod
 from dataclasses import dataclass, replace
-from typing import (TYPE_CHECKING, Dict, Iterable, List, Literal, NamedTuple,
-                    Optional, Tuple, Union)
+from typing import TYPE_CHECKING, Dict, Iterable, List, Literal, NamedTuple, Optional, Tuple, Union
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: `mamba_cache_manager.py` is in `legacy-files.txt`, so `.pre-commit-config.yaml` explicitly *excludes* it from `ruff-format` (line 1629) and hands it to `isort`/`yapf` at `line_length = 80` instead. The reflowed imports here are now 99/96/97 chars (lines 20, 35, 52) and isort will fold them back, which also leaves the file split between two styles — the new `commit_conv_window` signature is still yapf-wrapped. Could you drop the format-only hunks? Migrating this file to ruff would be a separate change via `scripts/legacy_utils.py gen-configs`.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 3]
```diff
@@ -2119,29 +2120,40 @@ def _load_trunk_params(
                 kda_tp_rank = layer.linear_attn._kda_tp_rank
                 break
 
+        COL = TensorParallelMode.COLUMN  # shards output rows (dim 0)
+        ROW = TensorParallelMode.ROW  # shards input columns (dim 1)
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py]: `linear_attn` is only constructed under `if self.is_kda:` (line 1528, mutually exclusive with `self_attn`), so `".linear_attn." in name` already implies the layer is KDA and `_layer_is_kda` is always true. The `is_kda_by_layer` list and the string-parsed layer index can go — a one-line comment stating the invariant carries the same information.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 4]
```diff
@@ -2198,9 +2210,32 @@ def load_param(name: str, param: torch.nn.Parameter):
...
+                if src.shape != param.shape:
+                    mode = ROW if name.endswith(".o_proj.weight") else COL
+                    src = load_weight_shard(src, kda_tp_size, kda_tp_rank, mode)
+            elif ".shared_experts." in name or ".mlp." in name:
+                # Shared-expert / dense-MLP: gate/up column, down row; TP factor
+                # from the shapes (the module is built at the per-rank dim). A
+                # subgroup smaller than model TP repeats, so the shard index is
+                # model tp_rank modulo the parameter's shard count.
+                if src.shape != param.shape:
+                    mode, axis = (ROW, 1) if name.endswith(".down_proj.weight") else (COL, 0)
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py]: Only `down_proj.weight` can reach this branch — `GatedMLP` always fuses gate/up, and `gate_up_proj.weight` returns early in the `_GATE_UP_FUSED_SUFFIX` block above, so the `else (COL, 0)` arm is unreachable. Suggest narrowing it to the `down_proj` ROW case it actually serves, and asserting `src.shape[1] % param.shape[1] == 0` before deriving `tp`: the old code's explicit divisibility guard is gone, so a non-divisible checkpoint now silently yields a floor `tp` and a misleading "shard result shape" error instead of the real shape mismatch.

File: tests/unittest/_torch/modules/kimi_kda/test_kimi_kda_verify_parity.py [hunk 5]
```diff
@@ -64,6 +64,9 @@ def __init__(self, slots, dim3, w, h, v, k, t_max, device):
         self.intermediate_ssm = torch.zeros(
             slots, t_max, h, v, k, dtype=torch.float32, device=device
         )
+        # This double drives the sequential-verify path, not the KDA
```
Comment (zhaoyangwang-nvidia) [inline tests/unittest/_torch/modules/kimi_kda/test_kimi_kda_verify_parity.py:67]: The four rewritten shard paths have no CI coverage — the test changes only add `has_kda_replay_caches` to the existing doubles, and `test_kimi_linear_checkpoint.py` only exercises the name plan, so the behavior-preservation claim rests entirely on a manual TP16 smoke. Could you add a shape-level unit test that runs synthetic checkpoint tensors (distinct non-zero values per position, so a wrong-rank slice is distinguishable from no slice) through `load_param` for a couple of `(kda_tp_size, rank)` combinations, covering the KDA COLUMN / `o_proj` ROW and shared-expert `down_proj` cases?

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 6]
```diff
@@ -2198,9 +2210,32 @@ def load_param(name: str, param: torch.nn.Parameter):
...
+                # KDA head-shard: the checkpoint is an exact kda_tp_size
+                # multiple on the head-major axis. Replicated projections
+                # (f_a/g_a, o_norm) already match and are copied as-is;
+                # o_proj shards its input columns (ROW), every other
+                # head-major projection its output rows (COLUMN). MLA
+                # projections are head-sharded by their own Linear modules
+                # (mla_head_shard_linears) in the shape-mismatch block below.
+                if src.shape != param.shape:
+                    mode = ROW if name.endswith(".o_proj.weight") else COL
+                    src = load_weight_shard(src, kda_tp_size, kda_tp_rank, mode)
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py]: These four `load_weight_shard` calls fall back to the default `device=torch.device('cpu')`, while the `load_shard` calls in the same function (lines 2182, 2241) pass `device=param.device`. It is a no-op today since `weights` are CPU safetensors slices, but passing `device=param.device` makes the assumption explicit and keeps the two loaders consistent.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 7]
```diff
@@ -2119,29 +2120,40 @@ def _load_trunk_params(
...
 
+        COL = TensorParallelMode.COLUMN  # shards output rows (dim 0)
+        ROW = TensorParallelMode.ROW  # shards input columns (dim 1)
+        is_kda_by_layer = [getattr(layer, "is_kda", False) for layer in self.model.layers]
+
+        def _layer_is_kda(name: str) -> bool:
+            idx = int(name.split("layers.", 1)[1].split(".", 1)[0])
+            return is_kda_by_layer[idx]
+
         def load_param(name: str, param: torch.nn.Parameter):
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py]: Still missing `-> None` here (as flagged above) — `load_param` writes `param.data` in place and returns nothing. Separately, the `if tp > 1 else 0` guards on lines 2148 and 2232 are redundant: `load_weight_shard` returns the whole tensor and ignores the rank when `tensor_parallel_size <= 1` (linear.py:177).

**Review by zhaoyangwang-nvidia — CHANGES_REQUESTED** [review]
This branch is 48 commits behind `main`, and `main` has since added Helix CP slicing for `v_b_proj` (`num_heads_tp_cp` / `cp_rank`) inside the same KV-B branch of `load_param` that this PR restructures. Please rebase before the next CI run and double-check that slicing survives the conflict — it only fires when `cp_size > 1`, so no smoke test would catch it going missing.

**Review by zhaoyangwang-nvidia — APPROVED** [review]
